client-core/src/service/key_service.rs

```diff
@@ -56,7 +92,152 @@ where
+
+                self.storage.set_secure(
+                    KEYSPACE,
+                    public_key.serialize(),
+                    private_key.serialize(),
+                    passphrase,
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    /// read value from db, if it's None, there value doesn't exist
+    pub fn read_value(&self, passphrase: &SecUtf8, key: &[u8]) -> Option<Vec<u8>> {
+        if let Ok(connected) = self.storage.get_secure(KEYSPACE_HD, key, passphrase) {
+            if let Some(value) = connected {
+                return Some(value.clone());
+            }
+        }
+        None
+    }
+
+    /// read number, if value doesn't exist, it returns default value
+    pub fn read_number(&self, passphrase: &SecUtf8, key: &[u8], default: u32) -> u32 {
+        if let Ok(connected) = self.storage.get_secure(KEYSPACE_HD, key, passphrase) {
+            if let Some(value) = connected {
+                return std::str::from_utf8(&value[..])
+                    .unwrap()
+                    .parse::<u32>()
+                    .unwrap();
+            }
+        }
+        default
+    }
+
+    /// write number to store, write number as string
+    /// writes hdwallet index, after making a new entry, index increases by 1
+    /// so address is generated in deterministic way.
+    pub fn write_number(&self, passphrase: &SecUtf8, key: &[u8], value: u32) {
+        let a = value.to_string();
+        let b = a.as_bytes();
+        self.storage
+            .set_secure(KEYSPACE_HD, key, b.to_vec(), passphrase)
+            .unwrap();
+    }
+
+    /// m / purpose' / coin_type' / account' / change / address_index
+    /// account: donation, savings, common expense
+    /// change: 0: external, 1: internal
+    /// Generates a new public-private keypair
+    pub fn generate_keypair_hd(
+        &self,
+        name: &str,           // wallet name
+        passphrase: &SecUtf8, // wallet pass phrase
+        is_staking: bool,     // kind of address
+    ) -> Result<(PublicKey, PrivateKey)> {
+        let seed_bytes = self.storage.get_secure(KEYSPACE_HD, name, passphrase)?;
+        let mut index = if is_staking {
+            self.read_number(passphrase, format!("staking_{}", name).as_bytes(), 0)
+        } else {
+            self.read_number(passphrase, format!("transfer_{}", name).as_bytes(), 0)
+        };
+        debug!("hdwallet index={}", index);
+        let cointype = get_bip44_coin_type();
+        println!("coin type={}", cointype);
+        let account = if is_staking { 1 } else { 0 };
+        let extended = ExtendedPrivKey::derive(
+            &seed_bytes.unwrap(),
+            format!("m/44'/{}'/{}'/0/{}", cointype, account, index).as_str(),
+        )
+        .unwrap();
+        let secret_key_bytes = extended.secret();
+        debug!("hdwallet save index={}", index);
+        let private_key = PrivateKey::deserialize_from(&secret_key_bytes).unwrap();
```

Comment:
Can you handle error instead of `unwrap`ing? Also, zeroize `secret_key_bytes`.

Reply:
ok